tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,286 @@ static bool HasCanonicalDimensionNumbers(
+                                : result_ty.getRank();
+      size_t batch_rank = operand_rank - (pre_elip + postElip);
+      // Add ("0",...,"N") where N is rank of batch into the loop
+      for (int batch_ind = 0; batch_ind < batch_rank; batch_ind++)
+        loop_dim.push_back(std::to_string(batch_ind));
+      // Add the dimension after ellipsis into the loop
+      for (int post_ch_ind = pre_elip + ellipsis.size();
+           post_ch_ind < loop.value().size(); post_ch_ind++)
+        loop_dim.push_back(std::string(1, loop.value()[post_ch_ind]));
+      batch_rank_vec.push_back(batch_rank);
+      all_loop_dim.push_back(std::move(loop_dim));
+    }
+    // Check that all ellipsis represent same rank of batch
+    if (batch_rank_vec.size() > 1) {
+      if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                      batch_rank_vec.begin())) {
+        return rewriter.notifyMatchFailure(
+            op, "Invalid Elipsis(ellipsis) within Einsum config!");
+      }
+    }
+
+    // Find all unique indices in the input and output
+    llvm::SmallSetVector<StringRef, 4> input_ind;
+    llvm::SmallSetVector<StringRef, 4> output_ind;
+    int num_of_input_operands = args.size();
+    for (const auto& loops_it : llvm::enumerate(all_loop_dim)) {
+      for (const StringRef& dim : loops_it.value()) {
+        if (loops_it.index() < max_operand_sz - 1) {
+          input_ind.insert(dim);
+        } else {
+          output_ind.insert(dim);
+        }
+      }
+    }
+
+    // Check for contraction/summation indices
+    llvm::SmallSetVector<StringRef, 4> reduction_axe =
+        FindSummationAxes(input_ind, output_ind);
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+    ValueRange inputs = args;
+
+    // Setting up Initial buffer for Output Tensor
+    Value output;
+    auto dyn_sizes = ExtractDynamicEinsumSizes(
+        rewriter, loc, args.front(), args.back(), all_loop_dim[0],
+        all_loop_dim[1], all_loop_dim[2]);
+    output = GetInitTensor(rewriter, loc, result_ty, dyn_sizes);
+
+    // Create indexing maps.
+    // Create a 1:1 map from f:StrDimension -> affineDimension
+    int64_t nloops = input_ind.size();
+    DenseMap<StringRef, AffineExpr> str_affine_dim_umap;
+    for (auto it : llvm::enumerate(input_ind)) {
+      str_affine_dim_umap[it.value()] = rewriter.getAffineDimExpr(it.index());
+    }
+    // For each Operands, we generate the affine map for accessing pattern.
+    // Using knowledge from einsum_config/loop of each operand in string type,
+    // we can generate the equivalent information in affine type using
+    // "GetExprFromConfig". Then we can feed it into the affinemap builder
+    // where: map = mapBuilder(number_of_dims, num_constant, einsum_config,
+    // mlirCtx)
+    SmallVector<AffineMap, 4> maps;
+    for (const auto& loop_operand : all_loop_dim) {
+      AffineMap scalar_map = AffineMap::get(nloops, 0, rewriter.getContext());
+      auto inputExprs =
```

Comment:
style nit: input_exprs.

Maybe just make it `exprs`?

Reply:
done